File: js/systems/particleSystem.js
class ParticleSystem {
    constructor(game) {
        this.game = game;
        this.particles = new Set();
        this.emitters = new Map();
        
        // Carregar emissores predefinidos
        this.loadEmitters();
    }

    loadEmitters() {
        // Carregar configurações de emissores
        EmitterDatabase.forEach((config, id) => {
            this.emitters.set(id, new ParticleEmitter(config));
        });
    }

    createEmitter(config) {
        return new ParticleEmitter(config);
    }

    emit(emitterId, position, options = {}) {
        const emitter = this.emitters.get(emitterId);
        if (!emitter) return;
        
        // Criar partículas
        const particles = emitter.emit(position, options);
        particles.forEach(particle => this.particles.add(particle));
    }

    update(deltaTime) {
        // Atualizar partículas
        this.particles.forEach(particle => {
            particle.update(deltaTime);
            
            // Remover partículas mortas
            if (particle.isDead()) {
                this.particles.delete(particle);
            }
        });
    }

    draw(ctx) {
        this.particles.forEach(particle => {
            particle.draw(ctx);
        });
    }

    clear() {
        this.particles.clear();
    }
}

class ParticleEmitter {
    constructor(config) {
        this.config = {
            // Quantidade de partículas
            count: config.count || 1,
            spread: config.spread || 0,
            
            // Propriedades das partículas
            size: config.size || { min: 2, max: 4 },
            speed: config.speed || { min: 50, max: 100 },
            angle: config.angle || { min: 0, max: 360 },
            life: config.life || { min: 0.5, max: 1.5 },
            
            // Cores e alpha
            colors: config.colors || ['#ff0000'],
            alpha: config.alpha || { start: 1, end: 0 },
            
            // Física
            gravity: config.gravity || 0,
            friction: config.friction || 0,
            
            // Comportamento
            behavior: config.behavior || 'normal',
            
            // Forma
            shape: config.shape || 'circle',
            
            // Efeitos
            effects: config.effects || {}
        };
    }

    emit(position, options = {}) {
        const particles = [];
        const count = options.count || this.config.count;
        
        for (let i = 0; i < count; i++) {
            // Criar partícula com propriedades aleatórias
            const particle = this.createParticle(position, options);
            particles.push(particle);
        }
        
        return particles;
    }

    createParticle(position, options) {
        // Posição inicial
        const spread = options.spread || this.config.spread;
        const x = position.x + (Math.random() - 0.5) * spread;
        const y = position.y + (Math.random() - 0.5) * spread;
        
        // Tamanho
        const size = this.randomRange(this.config.size);
        
        // Velocidade e direção
        const speed = this.randomRange(this.config.speed);
        const angle = this.randomRange(this.config.angle) * Math.PI / 180;
        const velocity = {
            x: Math.cos(angle) * speed,
            y: Math.sin(angle) * speed
        };
        
        // Tempo de vida
        const life = this.randomRange(this.config.life);
        
        // Cor
        const color = this.config.colors[
            Math.floor(Math.random() * this.config.colors.length)
        ];
        
        return new Particle({
            x, y,
            size,
            velocity,
            life,
            color,
            alpha: { ...this.config.alpha },
            gravity: this.config.gravity,
            friction: this.config.friction,
            behavior: this.config.behavior,
            shape: this.config.shape,
            effects: { ...this.config.effects }
        });
    }

    randomRange(range) {
        if (typeof range === 'number') return range;
        return range.min + Math.random() * (range.max - range.min);
    }
}

class Particle {
    constructor(config) {
        // Posição
        this.x = config.x;
        this.y = config.y;
        
        // Física
        this.velocity = config.velocity;
        this.gravity = config.gravity;
        this.friction = config.friction;
        
        // Aparência
        this.size = config.size;
        this.color = config.color;
        this.alpha = config.alpha;
        this.shape = config.shape;
        
        // Tempo de vida
        this.life = config.life;
        this.maxLife = config.life;
        
        // Comportamento
        this.behavior = config.behavior;
        
        // Efeitos
        this.effects = config.effects;
        this.initEffects();
    }

    initEffects() {
        Object.entries(this.effects).forEach(([key, effect]) => {
            if (typeof effect.current === 'undefined') {
                effect.current = effect.start;
                effect.delta = (effect.end - effect.start) / (this.maxLife * 1000);
            }
        });
    }

    update(deltaTime) {
        // Atualizar tempo de vida
        this.life -= deltaTime / 1000;
        
        // Atualizar física
        this.updatePhysics(deltaTime);
        
        // Atualizar efeitos
        this.updateEffects(deltaTime);
        
        // Atualizar comportamento específico
        this.updateBehavior(deltaTime);
    }

    updatePhysics(deltaTime) {
        // Aplicar gravidade
        this.velocity.y += this.gravity * deltaTime / 1000;
        
        // Aplicar fricção
        this.velocity.x *= (1 - this.friction * deltaTime / 1000);
        this.velocity.y *= (1 - this.friction * deltaTime / 1000);
        
        // Atualizar posição
        this.x += this.velocity.x * deltaTime / 1000;
        this.y += this.velocity.y * deltaTime / 1000;
    }

    updateEffects(deltaTime) {
        // Atualizar alpha
        if (this.alpha) {
            const progress = 1 - (this.life / this.maxLife);
            this.alpha.current = this.alpha.start + 
                (this.alpha.end - this.alpha.start) * progress;
        }
        
        // Atualizar outros efeitos
        Object.values(this.effects).forEach(effect => {
            effect.current += effect.delta * deltaTime;
        });
    }

    updateBehavior(deltaTime) {
        switch (this.behavior) {
            case 'spiral':
                this.updateSpiralBehavior(deltaTime);
                break;
            case 'wave':
                this.updateWaveBehavior(deltaTime);
                break;
            case 'attract':
                this.updateAttractBehavior(deltaTime);
                break;
        }
    }

    updateSpiralBehavior(deltaTime) {
        const angle = (this.maxLife - this.life) * 5;
        const radius = 50 * (this.life / this.maxLife);
        
        this.velocity.x = Math.cos(angle) * radius - this.x;
        this.velocity.y = Math.sin(angle) * radius - this.y;
    }

    updateWaveBehavior(deltaTime) {
        const frequency = 5;
        const amplitude = 20;
        
        this.velocity.y = Math.sin((this.maxLife - this.life) * frequency) * 
            amplitude * deltaTime / 1000;
    }

    updateAttractBehavior(deltaTime) {
        const target = { x: 0, y: 0 };
        const strength = 500;
        
        const dx = target.x - this.x;
        const dy = target.y - this.y;
        const distance = Math.sqrt(dx * dx + dy * dy);
        
        if (distance > 0) {
            this.velocity.x += (dx / distance) * strength * deltaTime / 1000;
            this.velocity.y += (dy / distance) * strength * deltaTime / 1000;
        }
    }

    draw(ctx) {
        ctx.save();
        
        // Aplicar alpha
        ctx.globalAlpha = this.alpha.current;
        
        // Aplicar cor
        ctx.fillStyle = this.color;
        
        // Desenhar forma
        switch (this.shape) {
            case 'circle':
                this.drawCircle(ctx);
                break;
            case 'square':
                this.drawSquare(ctx);
                break;
            case 'star':
                this.drawStar(ctx);
                break;
        }
        
        ctx.restore();
    }

    drawCircle(ctx) {
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.size / 2, 0, Math.PI * 2);
        ctx.fill();
    }

    drawSquare(ctx) {
        ctx.fillRect(
            this.x - this.size / 2,
            this.y - this.size / 2,
            this.size,
            this.size
        );
    }

    drawStar(ctx) {
        const spikes = 5;
        const outerRadius = this.size / 2;
        const innerRadius = outerRadius / 2;
        
        ctx.beginPath();
        ctx.translate(this.x, this.y);
        ctx.rotate(Math.PI / 2);
        
        for (let i = 0; i < spikes * 2; i++) {
            const radius = i % 2 === 0 ? outerRadius : innerRadius;
            const angle = (Math.PI * i) / spikes;
            
            if (i === 0) {
                ctx.moveTo(radius, 0);
            } else {
                ctx.lineTo(
                    radius * Math.cos(angle),
                    radius * Math.sin(angle)
                );
            }
        }
        
        ctx.closePath();
        ctx.fill();
    }

    isDead() {
        return this.life <= 0;
    }
}

// Banco de Dados de Emissores
const EmitterDatabase = new Map([
    ['hit_impact', {
        count: 10,
        spread: 20,
        size: { min: 3, max: 8 },
        speed: { min: 100, max: 200 },
        angle: { min: 0, max: 360 },
        life: { min: 0.3, max: 0.6 },
        colors: ['#ff0000', '#ff6666'],
        alpha: { start: 1, end: 0 },
        gravity: 300,
        friction: 0.5
    }],
    
    ['heal_effect', {
        count: 15,
        spread: 30,
        size: { min: 4, max: 8 },
        speed: { min: 50, max: 100 },
        angle: { min: -90, max: -270 },
        life: { min: 0.8, max: 1.2 },
        colors: ['#00ff00', '#66ff66'],
        alpha: { start: 1, end: 0 },
        gravity: -100,
        behavior: 'spiral'
    }],
    
    ['level_up', {
        count: 30,
        spread: 50,
        size: { min: 5, max: 10 },
        speed: { min: 150, max: 300 },
        angle: { min: 0, max: 360 },
        life: { min: 1, max: 1.5 },
        colors: ['#ffff00', '#ffdd00', '#ffd700'],
        alpha: { start: 1, end: 0 },
        gravity: -50,
        behavior: 'wave',
        shape: 'star'
    }]
]);

File: src/systems/particle_system.py
from typing import Dict, List, Optional, Tuple
import pygame
import random
import math

class Particle:
    def __init__(self, x: float, y: float, velocity: Tuple[float, float], 
                 color: Tuple[int, int, int], size: int, lifetime: float,
                 gravity: float = 0):
        self.x = x
        self.y = y
        self.velocity = list(velocity)
        self.color = color
        self.size = size
        self.lifetime = lifetime
        self.time_left = lifetime
        self.gravity = gravity
        self.alpha = 255
        
    def update(self, delta_time: float):
        """Atualiza a partícula."""
        self.time_left -= delta_time
        
        # Atualiza posição
        self.x += self.velocity[0] * delta_time
        self.y += self.velocity[1] * delta_time
        
        # Aplica gravidade
        self.velocity[1] += self.gravity * delta_time
        
        # Atualiza transparência
        self.alpha = int((self.time_left / self.lifetime) * 255)
        
    def is_alive(self) -> bool:
        """Verifica se a partícula ainda está viva."""
        return self.time_left > 0
        
    def draw(self, screen: pygame.Surface, camera_x: int = 0, camera_y: int = 0):
        """Desenha a partícula."""
        if self.alpha <= 0:
            return
            
        # Cria uma superfície para a partícula com alpha
        particle_surface = pygame.Surface((self.size, self.size), pygame.SRCALPHA)
        color_with_alpha = (*self.color, self.alpha)
        pygame.draw.circle(particle_surface, color_with_alpha, 
                         (self.size // 2, self.size // 2), self.size // 2)
                         
        # Desenha na tela
        screen.blit(particle_surface, 
                   (self.x - camera_x - self.size // 2, 
                    self.y - camera_y - self.size // 2))

class ParticleSystem:
    def __init__(self):
        self.particles: List[Particle] = []
        
    def create_particle(self, x: float, y: float, 
                       velocity: Optional[Tuple[float, float]] = None,
                       color: Optional[Tuple[int, int, int]] = None,
                       size: Optional[int] = None,
                       lifetime: Optional[float] = None,
                       gravity: float = 0):
        """Cria uma nova partícula."""
        # Valores padrão
        if velocity is None:
            angle = random.uniform(0, 2 * math.pi)
            speed = random.uniform(50, 100)
            velocity = (math.cos(angle) * speed, math.sin(angle) * speed)
            
        if color is None:
            color = (random.randint(0, 255), 
                    random.randint(0, 255), 
                    random.randint(0, 255))
                    
        if size is None:
            size = random.randint(2, 6)
            
        if lifetime is None:
            lifetime = random.uniform(0.5, 2.0)
            
        particle = Particle(x, y, velocity, color, size, lifetime, gravity)
        self.particles.append(particle)
        
    def create_explosion(self, x: float, y: float, particle_count: int = 20):
        """Cria um efeito de explosão."""
        for _ in range(particle_count):
            angle = random.uniform(0, 2 * math.pi)
            speed = random.uniform(100, 200)
            velocity = (math.cos(angle) * speed, math.sin(angle) * speed)
            
            color = (random.randint(200, 255),  # Vermelho
                    random.randint(0, 100),     # Verde
                    0)                          # Azul
                    
            self.create_particle(x, y, velocity, color, 
                               random.randint(4, 8), 
                               random.uniform(0.5, 1.0),
                               gravity=200)
                               
    def create_sparkle(self, x: float, y: float, particle_count: int = 5):
        """Cria um efeito de brilho."""
        for _ in range(particle_count):
            angle = random.uniform(0, 2 * math.pi)
            speed = random.uniform(20, 50)
            velocity = (math.cos(angle) * speed, math.sin(angle) * speed)
            
            color = (255, 255, random.randint(200, 255))  # Branco/Amarelo
            
            self.create_particle(x, y, velocity, color,
                               random.randint(2, 4),
                               random.uniform(0.2, 0.5))
                               
    def create_trail(self, x: float, y: float, direction: Tuple[float, float],
                    color: Tuple[int, int, int], particle_count: int = 1):
        """Cria um efeito de rastro."""
        for _ in range(particle_count):
            offset_x = random.uniform(-5, 5)
            offset_y = random.uniform(-5, 5)
            
            # Velocidade oposta à direção do movimento
            speed = random.uniform(10, 30)
            velocity = (-direction[0] * speed + random.uniform(-10, 10),
                      -direction[1] * speed + random.uniform(-10, 10))
                      
            self.create_particle(x + offset_x, y + offset_y,
                               velocity, color,
                               random.randint(2, 4),
                               random.uniform(0.2, 0.5))
                               
    def update(self, delta_time: float):
        """Atualiza todas as partículas."""
        # Atualiza e remove partículas mortas
        self.particles = [particle for particle in self.particles 
                         if particle.is_alive()]
                         
        for particle in self.particles:
            particle.update(delta_time)
            
    def draw(self, screen: pygame.Surface, camera_x: int = 0, camera_y: int = 0):
        """Desenha todas as partículas."""
        for particle in self.particles:
            particle.draw(screen, camera_x, camera_y)
